File: c++/sorting/quick-sort.cpp
#include <iostream>
#define max 100

using namespace std;

int quicksort(int arr[], int low, int high)
{
    int i, j, pivot, temp;

    if (low < high)
    {
        pivot = low;
        i = low;
        j = high;

        while (i < j)
        {
            while (arr[i] <= arr[pivot])
            {
                i++;
            }

            while (arr[j] > arr[pivot])
            {
                j--;
            }

            if (i < j)
            {
                temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        temp = arr[pivot];
        arr[pivot] = arr[j];
        arr[j] = temp;

        quicksort(arr, low, j - 1);
        quicksort(arr, j + 1, high);
    }
}

int main()
{

    int arr[max], n;

    cout << "enter how many elements : " << endl;
    cin >> n;

    cout << "enter elements one by one : " << endl;
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }

    quicksort(arr, 0, n - 1);

    cout << "finally sorted array is : " << endl;

    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }

    cout << endl;
}

File: python/sorting/quicksort.py
arr = [5,3,2,1, 60]

def quicksort(arr, low, high):
    if low < high:
        partition_pos = partition(arr, low, high)
        quicksort(arr, low, partition_pos-1)
        quicksort(arr, partition_pos+1, high)
        
def partition(arr, l, h):
    pivot = h
    i = l
    j = h-1
    
    while i < j:
        while i < h and arr[i] <= arr[pivot]:
            i += 1
        while j > l and arr[j] >= arr[pivot]:
            j -= 1
    
        if i < j:
            arr[i], arr[j] = arr[j], arr[i]
    
    if arr[i] > arr[pivot]:
        arr[i], arr[pivot] = arr[pivot], arr[i]
    
    return i
    
quicksort(arr, 0, len(arr)-1)
print(arr)
        
